File: data_processing/scripts/areaMaker.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""
@author: Latour Rodolphe
"""

import rospy
import message_filters
from data_processing.msg import ObjectMsg
from data_processing.msg import ObjectsMsg
from yolo_madnet.msg import PointsMsg
from nav_msgs.msg import Odometry

from object_creator import object_creator
import numpy as np
from colorama import Fore, Style
from pyquaternion import Quaternion
import time



global DECAYING_TIME
DECAYING_TIME = 5

class areaMaker:
    """
    Description:
    ============
    This script is built to filter the points taken from /object/detected topic. It
    creates areas that have in attributes a name depending on the class of the
    detected object, a location, a rotation a scale, a creation time, a deadline,
    and an existence probability.

    The goal of an existence area is to classify added points to avoid overfeedings
    of point objects. It means that with recurrence of detection, the program gives
    the ability to create a coercition from previous added point in a specific
    area. Also, from the score of the detection and the recurrence, we can set an
    existence probability that can be used to avoid error of detections.
    """
    def __init__(self):

        print('Initializing node...')
        rospy.init_node('areamaker_node')
        subObj = message_filters.Subscriber('/object/detected', PointsMsg)
        subcampos = message_filters.Subscriber('/t265/odom/sample', Odometry)
        ats = message_filters.ApproximateTimeSynchronizer([subObj, subcampos], queue_size=2, slop=0.01)
        ats.registerCallback(self.process)
        self.obj_list = []
        self.pub = rospy.Publisher('/object/position/3D', ObjectsMsg, queue_size=1)
        self.time =time.time()
        rospy.spin()


    def transpose_to_global(self, point, cam_point, quaternion):
        """
        Description:
        ============
        This function is used to transpose the position of the point into global
        referential based on the tracking camera position.

        Input:
        ------
        - point: The position of the point (x,y,z) based on the camera referential
        - cam_point: The position of the tracking camera T265
        - quatertion: The rotation of the tracking camera T265

        Output:
        -------
        - point: The position of the input point set in global referential
        """
        # Perform rotation based on T265 quaternion
        point = quaternion.rotate(point)
        # Convert to numpy matrix format
        point = np.matrix([[point[0],
                            point[1],
                            point[2],
                            1]])
        cam_point = np.matrix([[1,0,0,cam_point[0][0]],
                    [0,1,0,cam_point[1][0]],
                    [0,0,1,cam_point[2][0]],
                    [0,0,0,1]])
        point = cam_point * point.T
        return point[:3]

    def process(self, pointsMsg, cameraPosMsg):
        """
        Description:
        ============
        Main function of the class.

        Input:
        ------
        - pointMsg: A list of all detected object. Each element is built
        that way:
            * Vector3 position: (x, y, z) position based of camera referential

            * Vector3 scale: 3D box scale (x, y, z) based on the bounding box.

            * string obj_class: The class of the object

            * float32 score: the detection score of the object

        - cameraPosMsg: All the data provided by the Realsense T265 camera.
        The used data here will be position (x,y,z) in meter and the orientation
        (w, i, j, k) given in quaternion.

        Output:
        -------
        - Objects: A list of object containing specific data.
            * Vector3 center:
             -- x: the x position in space (relative to the global position)
             -- y: the y position in space (relative to the global position)
             -- z: the z position in space (relative to the global position)

            * Quaternion rotation: The rotation of the object

            * Vector3 scale
             -- scale_x: the scale in x of the box based on the bounding box
             -- scale_y: the scale in y of the box based on the bounding box
             -- scale_z: the scale in z of the box based on the bounding box

            * float64 creation_time: the creation time of the object

            * float64 last_detection_time: the last detection time of the object

            * string obj_class: The class of the object

            * float32 score: the detection score of the object

            * int32 ID: The unique ID of the object
        """
        print("\n__________________________________\n")
        start = time.time()

        objects = ObjectsMsg()
        redondance = 0

        for pt in pointsMsg.point:

            # Get the point position, class and score
            # Convert point referential to camera referential
            x = pt.position.z
            y = pt.position.x
            z = pt.position.y
            print(z)
            if z > 10:
                continue
            local_point = np.matrix([[x],[y],[z]])
            _class = pt.obj_class
            score = pt.score
            id = pt.id
            scale = (pt.scale.z, pt.scale.x, pt.scale.y)


            # Get the camera position (euler vector) and rotation (quaternion)
            cam_x = cameraPosMsg.pose.pose.position.x
            cam_y = cameraPosMsg.pose.pose.position.y
            cam_z = cameraPosMsg.pose.pose.position.z
            cam_point = np.matrix([[cam_x], [cam_y], [cam_z]])

            # If tracking camera is looking ahead
            # cam_rx = cameraPosMsg.pose.pose.orientation.x
            # cam_ry = cameraPosMsg.pose.pose.orientation.y
            # cam_rz = cameraPosMsg.pose.pose.orientation.z
            # cam_rw = cameraPosMsg.pose.pose.orientation.w

            # Else if tracking camera is looking behind
            cam_rx =   cameraPosMsg.pose.pose.orientation.z
            cam_ry =   cameraPosMsg.pose.pose.orientation.w
            cam_rz = - cameraPosMsg.pose.pose.orientation.x
            cam_rw =   cameraPosMsg.pose.pose.orientation.y
            quaternion = Quaternion(cam_rw, cam_rx, cam_ry, cam_rz)


            # Print the point position in the camera and global referentials
            print(Fore.BLUE + "\nPoint position: ")
            print("#-----------REFENTIALS-----------#" + Style.RESET_ALL)
            print("Camera referential: \n{}".format(local_point))

            global_point = self.transpose_to_global(local_point, cam_point, quaternion)
            print("Global referential: \n{}".format(global_point))
            print(Fore.BLUE + "#-----------ALL OBJECTS-----------#")


            # If no object exists, we create a new mesh at the given position of the added point
            if len(self.obj_list) == 0:
                self.obj_list.append(object_creator(global_point, cam_point, scale, quaternion, _class, score, id))


            # Run through each existence area to check if the added point is inside and do processing
            for item in self.obj_list:
                # If the class of the item doesn't match with detection, skip

                if not _class == item.get_class():
                    print(_class, item.get_class())
                    lock_obj_creator = False
                    continue

                else:
                    print(Fore.BLUE + "#-----------Item {}-----------#".format(self.obj_list.index(item)) + Style.RESET_ALL)
                    print("Object center: \n{}".format(item.get_center()))
                    print("camera position: \n{}".format((cam_x,cam_y,cam_z)))
                    print("Object scale: \n{}".format(item.get_scale()))
                    print("Object class: \n{}".format(item.get_class()))
                    print("Object score: \n{}".format(round(item.get_score(),2)))
                    print("Object ID: \n{}".format(item.get_ID()))
                    iou = item.iou_3D(scale, global_point)
                    if iou >= 0.1: # and item.is_inside(global_point):
                        print(Fore.GREEN + "IoU: \n{}".format(iou) + Style.RESET_ALL)


                        # Perform calibration
                        item.calibrate(global_point, cam_point, quaternion, score, scale)
                        print("Iteration: {}".format(item.get_iteration()))

                        # Say that no object must be created
                        lock_obj_creator = True

                        # get all data Publish the position of the object
                        msg_center = item.get_center()
                        msg_scale = item.get_scale()
                        msg_rotation = item.get_quaternion()
                        msg_class = item.get_class()
                        msg_score = item.get_score()
                        msg_ID = item.get_ID()
                        msg_creation_time = item.get_creation_time()
                        msg_last_detection_time = item.get_last_detection_time()
                        now = rospy.get_rostime()
                        object = ObjectMsg()
                        object.header.stamp.secs = now.secs
                        object.header.stamp.nsecs = now.nsecs
                        object.center.x = msg_center[0]
                        object.center.y = msg_center[1]
                        object.center.z = msg_center[2]
                        object.scale.x = msg_scale[0]
                        object.scale.y = msg_scale[1]
                        object.scale.z = msg_scale[2]
                        object.rotation.w = msg_rotation[0]
                        object.rotation.x = msg_rotation[1]
                        object.rotation.y = msg_rotation[2]
                        object.rotation.z = msg_rotation[3]
                        object.creation_time = msg_creation_time
                        object.last_detection_time = msg_last_detection_time
                        object.obj_class = msg_class
                        object.score = msg_score
                        object.ID = msg_ID
                        objects.object.append(object)
                        break

                    # If the box doesn't match the existence area...
                    # let the ability to create a new one
                    else:
                        print(Fore.RED + "IoU: \n{}".format(iou) + Style.RESET_ALL)
                        lock_obj_creator = False
                    print("Creation time: \n{}".format(round(item.get_creation_time() - self.time,3)))

                    try:
                        dt = time.time() - item.get_last_detection_time()
                        print("Last detection: \n{}".format(round(dt,6)))
                        # If an object haven't been detected since a specific time, remove it.
                        if dt > DECAYING_TIME:
                            self.obj_list.remove(item)


                    # Avoid synchronization error (rare normally)
                    except AttributeError:
                        print("ERROR: Cannot get last detection time.")


            # Create a new object (a limit is added for the test)
            if not lock_obj_creator:# and len(self.obj_list) < 5:
                self.obj_list.append(object_creator(global_point, cam_point, scale, quaternion, _class, score, id))
            print("Processing time: {} ms".format(round((time.time()-start)*1000,3)))

        # Print the number of detected objects in the environment
        now = rospy.get_rostime()
        objects.header.stamp.secs = now.secs
        objects.header.stamp.nsecs = now.nsecs
        self.pub.publish(objects)
        print("Detected objects: {}".format(len(self.obj_list)))

if __name__ == '__main__':
    try:
        areaMaker()
    except rospy.ROSInterruptException:
        rospy.logerr('Could not start area maker node.')

File: data_processing/src/areaMaker.cpp
#include "../include/areaMaker.h"

using namespace yolo_madnet;
using namespace std;
using namespace message_filters;

/**
The class is a c++ version of areaMaker.py. It is still under construction and
might be used to replace the python version for a more efficient one.
**/

areaMaker::areaMaker()
{
  subObj.subscribe(nh, "/object/detected", 1);
  subCamPos.subscribe(nh, "/t265/odom/sample", 1);
  sync.reset(new Sync(_SyncPolicy(10), subObj, subCamPos));
  sync->registerCallback(boost::bind(&areaMaker::callback, this, _1, _2));
  pub = nh.advertise<data_processing::ObjectsMsg>("/object/position/3D", 1);
}

/*
This method is used to perform the rotation of the vector v by the Quaternion q
and the result is a rotated vector vprime.
*/
void areaMaker::rotateVectorByQuaternion(const struct Vector3& v, const struct Quaternion& q, struct Vector3& vprime)
{
    float num12 = q.x + q.x;
    float num2 = q.y + q.y;
    float num = q.z + q.z;
    float num11 = q.w * num12;
    float num10 = q.w * num2;
    float num9 = q.w * num;
    float num8 = q.x * num12;
    float num7 = q.x * num2;
    float num6 = q.x * num;
    float num5 = q.y * num2;
    float num4 = q.y * num;
    float num3 = q.z * num;
    float num15 = ((v.x * ((1.0f - num5) - num3)) + (v.y * (num7 - num9))) + (v.z * (num6 + num10));
    float num14 = ((v.x * (num7 + num9)) + (v.y * ((1.0f - num8) - num3))) + (v.z * (num4 - num11));
    float num13 = ((v.x * (num6 - num10)) + (v.y * (num4 + num11))) + (v.z * ((1.0f - num8) - num5));
    vprime.x = num15;
    vprime.y = num14;
    vprime.z = num13;
}
/***
  Description:
  ============
  This function is used to transpose the position of the point into global
  referential based on the tracking camera position.

  Input:
  ------
  - point: The position of the point (x,y,z) based on the camera referential
  - cam_point: The position of the tracking camera T265
  - quatertion: The rotation of the tracking camera T265

  Output:
  -------
  - point: The position of the input point set in global referential
***/
struct Vector3 areaMaker::transposeToGlobal(struct Vector3 point, struct Vector3 camPoint, struct Quaternion quaternion)
{
  struct Vector3 v{0,0,0};
  areaMaker::rotateVectorByQuaternion(point, quaternion, v);
  v.x += camPoint.x;
  v.y += camPoint.y;
  v.z += camPoint.z;
  return v;
}

/***
  Description:
  ============
  Main function of the class.

  Input:
  ------
  - pointMsg: A list of all detected object. Each element is built
  that way:
      * Vector3 position: (x, y, z) position based of camera referential

      * Vector3 scale: 3D box scale (x, y, z) based on the bounding box.

      * string obj_class: The class of the object

      * float32 score: the detection score of the object

  - cameraPosMsg: All the data provided by the Realsense T265 camera.
  The used data here will be position (x,y,z) in meter and the orientation
  (w, i, j, k) given in quaternion.

  Output:
  -------
  - Objects: A list of object containing specific data.
      * Vector3 center:
       -- x: the x position in space (relative to the global position)
       -- y: the y position in space (relative to the global position)
       -- z: the z position in space (relative to the global position)

      * Quaternion rotation: The rotation of the object

      * Vector3 scale
       -- scale_x: the scale in x of the box based on the bounding box
       -- scale_y: the scale in y of the box based on the bounding box
       -- scale_z: the scale in z of the box based on the bounding box

      * float64 creation_time: the creation time of the object

      * float64 last_detection_time: the last detection time of the object

      * string obj_class: The class of the object

      * float32 score: the detection score of the object

      * int32 ID: The unique ID of the object
***/
void areaMaker::callback(const PointsMsgConstPtr& pointsMsg, const nav_msgs::OdometryConstPtr& cameraPosMsg)
{
  data_processing::ObjectsMsg objects;
  bool lockObjCreator = true;
  float iou;

  for (size_t i = 0; i < pointsMsg->point.size(); i++)
  {
    PointMsg pt = pointsMsg->point[i];
    // Create a vector for local point
    struct Vector3 localPoint{pt.position.z, - pt.position.x, pt.position.y};

    // We don't want to care about object distance upper 8 meters
    if (localPoint.x > 8.0){
      continue;
    }

    // Get class, score and ID of the object
    string const _class = pt.obj_class;
    double const score = pt.score;
    int long const ID = pt.id;

    // Get the scale of the object's box
    // Convert scale referential to camera referential
    struct Vector3 const scale{pt.scale.z, pt.scale.x, pt.scale.y};

    // Get the camera position (euler vector) and rotation (quaternion)
    struct Vector3 const camPoint{cameraPosMsg->pose.pose.position.x,
                            cameraPosMsg->pose.pose.position.y,
                            cameraPosMsg->pose.pose.position.z};

    struct Quaternion const quaternion{cameraPosMsg->pose.pose.orientation.z,
                                  cameraPosMsg->pose.pose.orientation.w,
                                - cameraPosMsg->pose.pose.orientation.x,
                                  cameraPosMsg->pose.pose.orientation.y};

    // Transpose local_point into global referential
    struct Vector3 globalPoint = areaMaker::transposeToGlobal(localPoint, camPoint, quaternion);

    // If no object exists, we create a new mesh at the given position of the added point
    if (m_objList.size() == 0)
    {
      m_objList.push_back(ObjectCreator(globalPoint, camPoint, scale, quaternion, _class, score, ID));
    }

    // Run through each box created
    for (size_t i = 0; i < m_objList.size(); i++)
    {
      // If the class of the item doesn't match with detection, skip
      if (_class != m_objList.at(i).getClass())
      {
        lockObjCreator = false;
        continue;
      }

      else if (ID == m_objList.at(i).getID())
      {
        iou = m_objList.at(i).IoU3D(scale, globalPoint);
        // Perform calibration
        m_objList.at(i).calibrate(globalPoint, quaternion, score, scale);
        lockObjCreator = true;
        // Be careful!!!
        // The ID sent is NOT the ID provided by Sort (cause it is
        // unstable). So, we send for rviz the index of the item in
        //the item list because it is unique.
        objects.object.push_back(joinData(m_objList.at(i), i));
        break;
      }
      else
      {
        iou = m_objList.at(i).IoU3D(scale, globalPoint);
        if (iou > 0)
        {
          // Perform calibration
          m_objList.at(i).calibrate(globalPoint, quaternion, score, scale);
          // We change the ID of the object. The previous one might
          // has been lost by Sort for major cases.
          m_objList.at(i).setID(ID);

          // PROBLEM WITH ITERATION... Stay at '1', no incrementation.


          // Say that no object must be created
          lockObjCreator = true;
          objects.object.push_back(joinData(m_objList.at(i), i));
          break;
        }
        else
        {
          lockObjCreator = false;
        }
      }
      if (time(0) - m_objList.at(i).getLastDetectionTime() > 50)
      {
        m_objList.erase(m_objList.begin() + i);
      }
    }
    if (!lockObjCreator)
    {
      m_objList.push_back(ObjectCreator(globalPoint, camPoint, scale, quaternion, _class, score, ID));
    }
  }
  ros::Time now = ros::Time::now();
  objects.header.stamp = now;
  pub.publish(objects);
}

/***
  Description:
  ============
  Use this to set all data regarding the item in an object ROS message.
  It set header, position and rotation, class, existence probability
  score, id and time details on the message.

  Input:
  ------
  - item: an object from ObjectCreator class.
  - id: The id of the object. Not the Sort one to avoid misinformation.

  Output:
  -------
  - object: a ROS type message from ObjectMsg.msg. It is used to be
  appended in ROS type message from ObjectsMsg.msg.
***/
data_processing::ObjectMsg areaMaker::joinData(ObjectCreator item, int id)
{
  // get all data Publish the position of the object
  struct Vector3 msg_center = item.getCenter();
  struct Vector3 msg_scale = item.getScale();
  struct Quaternion msg_rotation = item.getQuaternion();
  string msg_class = item.getClass();
  float msg_score = item.getScore();
  int msg_ID = id;
  float msg_creation_time = item.getCreationTime();
  float msg_last_detection_time = item.getLastDetectionTime();
  ros::Time now = ros::Time::now();
  data_processing::ObjectMsg object;
  object.header.stamp = now;
  object.center.x = msg_center.x;
  object.center.y = msg_center.y;
  object.center.z = msg_center.z;
  object.scale.x = msg_scale.x;
  object.scale.y = msg_scale.y;
  object.scale.z = msg_scale.z;
  object.rotation.w = msg_rotation.w;
  object.rotation.x = msg_rotation.x;
  object.rotation.y = msg_rotation.y;
  object.rotation.z = msg_rotation.z;
  object.creation_time = msg_creation_time;
  object.last_detection_time = msg_last_detection_time;
  object.obj_class = msg_class;
  object.score = msg_score;
  object.ID = msg_ID;
  return object;
}

int main(int argc, char** argv){

  ros::init(argc, argv, "areaMaker_cpp_node");

  areaMaker am;

  ros::spin();

  return 0;
}
